pyro/infer/mcmc/api.py

```diff
@@ -152,13 +152,20 @@ def terminate(self, *args, **kwargs):
 
     def run(self, *args, **kwargs):
         logger = logging.getLogger("pyro.infer.mcmc")
-        progress_bar = ProgressBar(self.warmup_steps, self.num_samples, disable=self.disable_progbar)
-        logger = initialize_logger(logger, "", progress_bar)
-        hook_w_logging = _add_logging_hook(logger, progress_bar, self.hook)
-        for sample in _gen_samples(self.kernel, self.warmup_steps, self.num_samples, hook_w_logging,
-                                   *args, **kwargs):
-            yield sample, 0  # sample, chain_id (default=0)
-        progress_bar.close()
+        for i in range(self.num_chains):
+            if self.initial_params is not None:
+                initial_params = {k: v[i] for k, v in self.initial_params.items()}
+                self.kernel.initial_params = initial_params
+
+            progress_bar = ProgressBar(self.warmup_steps, self.num_samples, disable=self.disable_progbar)
+            logger = initialize_logger(logger, "", progress_bar)
+            hook_w_logging = _add_logging_hook(logger, progress_bar, self.hook)
+            for sample in _gen_samples(self.kernel, self.warmup_steps, self.num_samples, hook_w_logging,
+                                       i if self.num_chains > 1 else None,
+                                       *args, **kwargs):
+                yield sample, i  # sample, chain_id
+            self.kernel.cleanup()
```

Comment:
We do a lot of state setting in our kernel classes, and `kernel.cleanup()` has not been very thoroughly tested because we never needed to do this for either single / parallel multi-chaining. We will need that for sequential chaining though. If you find that to be an issue, it might be worth handling that separately. Happy to take a look too.

Reply:
I am also worried about it so I added `test_sequential_consistence` to verify `cleanup` is done properly. Some changes in adapter class clean up adapter too.